include/deal.II/base/function.h

```diff
@@ -347,135 +347,158 @@ class Function : public FunctionTime<Number>,
-  virtual void vector_value_list (const std::vector<Point<dim> > &points,
-                                  std::vector<Vector<Number> >   &values) const;
+    virtual void value_list (const std::vector<Point<dim> > &points,
+                             std::vector<Number>            &values,
+                             const unsigned int              component = 0) const;
 
-  virtual Tensor<1,dim, Number> gradient (const Point<dim> &p,
-                                          const unsigned int component = 0) const;
+    virtual void vector_value_list (const std::vector<Point<dim> > &points,
+                                    std::vector<Vector<Number> >   &values) const;
 
-  virtual void vector_gradient (const Point<dim>            &p,
-                                std::vector<Tensor<1,dim, Number> > &gradients) const;
+    virtual Tensor<1,dim, Number> gradient (const Point<dim> &p,
+                                            const unsigned int component = 0) const;
 
-  virtual void gradient_list (const std::vector<Point<dim> > &points,
-                              std::vector<Tensor<1,dim, Number> >    &gradients,
-                              const unsigned int              component = 0) const;
+    virtual void vector_gradient (const Point<dim>            &p,
+                                  std::vector<Tensor<1,dim, Number> > &gradients) const;
 
-  virtual void vector_gradient_list (const std::vector<Point<dim> >            &points,
-                                     std::vector<std::vector<Tensor<1,dim, Number> > > &gradients) const;
-};
+    virtual void gradient_list (const std::vector<Point<dim> > &points,
+                                std::vector<Tensor<1,dim, Number> >    &gradients,
+                                const unsigned int              component = 0) const;
+
+    virtual void vector_gradient_list (const std::vector<Point<dim> >            &points,
+                                       std::vector<std::vector<Tensor<1,dim, Number> > > &gradients) const;
+  };
 
 
 
+  /**
+   * Provide a function which always returns the constant values handed to the
+   * constructor.
+   *
+   * Obviously, the derivatives of this function are zero, which is why we derive
+   * this class from <tt>ZeroFunction</tt>: we then only have to overload the
+   * value functions, not all the derivatives. In some way, it would be more
+   * obvious to do the derivation in the opposite direction, i.e. let
+   * <tt>ZeroFunction</tt> be a more specialized version of
+   * <tt>ConstantFunction</tt>; however, this would be less efficient, since we
+   * could not make use of the fact that the function value of the
+   * <tt>ZeroFunction</tt> is known at compile time and need not be looked up
+   * somewhere in memory.
+   *
+   * @ingroup functions
+   * @author Wolfgang Bangerth, 1998, 1999, Lei Qiao, 2015
+   */
+  template <int dim, typename Number=double>
+  class ConstantFunction : public ZeroFunction<dim, Number>
```

Comment:
Yes, we should. We have some wonky class inheritances in deal.II that were made for pure convenience instead of to represent an is-a relationship. Here, the convenience was probably that `ZeroFunction::gradient()` already returned all zeros and so we didn't have to re-implement the function.

I think what we should actually do is the opposite: let `ZeroFunction` derive from `ConstantFunction` and only re-implement the `value()` function.